crates/client/settlement_client/src/starknet/mod.rs

```diff
@@ -0,0 +1,624 @@
+    }
+
+    async fn get_last_state_root(&self) -> anyhow::Result<Felt> {
+        // State Root index in call response : 0
+        Ok(self.get_state_call().await?[0])
+    }
+
+    async fn get_last_verified_block_hash(&self) -> anyhow::Result<Felt> {
+        // Block Hash index in call response : 2
+        Ok(self.get_state_call().await?[2])
+    }
+
+    async fn get_initial_state(&self) -> anyhow::Result<StateUpdate> {
+        let block_number = self.get_last_verified_block_number().await?;
+        let block_hash = self.get_last_verified_block_hash().await?;
+        let global_root = self.get_last_state_root().await?;
+
+        Ok(StateUpdate { global_root, block_number, block_hash })
+    }
+
+    async fn listen_for_update_state_events(
+        &self,
+        backend: Arc<MadaraBackend>,
+        mut ctx: ServiceContext,
+    ) -> anyhow::Result<()> {
+        loop {
+            let events_response = ctx.run_until_cancelled(self.get_events(
+                BlockId::Number(self.get_latest_block_number().await?),
+                BlockId::Number(self.get_latest_block_number().await?),
+                self.l2_core_contract,
+                vec![get_selector_from_name("LogStateUpdate")?],
+            ));
+
+            match events_response.await {
+                Some(Ok(emitted_events)) => {
+                    if let Some(event) = emitted_events.last() {
+                        let data = event; // Create a longer-lived binding
+                        let formatted_event = StateUpdate {
+                            block_number: data.data[1].to_u64().expect("Unable to parse Felt result into u64"),
+                            global_root: data.data[0],
+                            block_hash: data.data[2],
+                        };
+                        update_l1(&backend, formatted_event, self.get_l1_block_metrics())?;
+                    }
+                }
+                Some(Err(e)) => {
+                    error!("Error processing event: {:?}", e);
+                }
+                None => {
+                    trace!("Starknet Client : No event found");
+                }
+            }
+
+            sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    async fn listen_for_messaging_events(
+        &self,
+        backend: Arc<MadaraBackend>,
+        mut ctx: ServiceContext,
+        last_synced_event_block: LastSyncedEventBlock,
+        chain_id: ChainId,
+        mempool: Arc<Mempool>,
+    ) -> anyhow::Result<()> {
+        let processor = self.message_processor(backend.clone(), chain_id, mempool.clone());
+        let mut sync_flag = false;
+
+        loop {
+            let latest_block_number = self.get_latest_block_number().await?;
+            let starting_block = if !sync_flag {
+                last_synced_event_block.block_number
+            } else if latest_block_number < 100 {
+                0
+            } else {
+                latest_block_number - 100
```

Comment:
instead of looping over last 100 blocks every time, I think we can keep state in memory here. the stream implementation i linked to allows for that